### ISSUE TITLE
TfLite toco: fail kTensorflowMerge

### ISSUE THREAD

**Issue by Kayoku** [tensorflow/tensorflow#18731]
### System information

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 18.04 Beta
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
Master
- **Python version**: 
3.6
- **Bazel version (if compiling from source)**:
0.10.0
- **GCC/Compiler version (if compiling from source)**:
gcc 4.8.5
- **CUDA/cuDNN version**:
9.1 / 7.0.5
- **GPU model and memory**:
GTX 1070 Ti 6go & 16Go RAM
- **Exact command to reproduce**:
bazel-bin/tensorflow/contrib/lite/toco/toco 
--input_file=frozen_graph.pb 
--input_format=TENSORFLOW_GRAPHDEF 
--output_format=TFLITE 
--output_file=/tmp/test.tflite 
--inference_input_type=FLOAT
--input_arrays=input_image
--output_arrays=output_corner,output_mask
--input_shapes=1,512,512,3

### Describe the problem

I'm working on a FCN. I want to export my frozen graph to TfLite graph but it's not working. I'm wondering why, but I guess it's just due to the lack of maturity of TfLite (and probably some operator compatibility). In my case, it's about the operator "Merge". Someone can explain me if there is a way/trick to make it work ?

### Source code / logs

```
2018-04-20 12:57:33.287154: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1268] Converting unsupported operation: LogicalOr
2018-04-20 12:57:33.302765: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 518 operators, 965 arrays (0 quantized)
2018-04-20 12:57:33.312501: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 477 operators, 902 arrays (0 quantized)
2018-04-20 12:57:33.325133: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 477 operators, 902 arrays (0 quantized)
2018-04-20 12:57:33.325335: F tensorflow/contrib/lite/toco/graph_transformations/resolve_tensorflow_switch.cc:95] Check failed: other_op->type == OperatorType::kTensorFlowMerge 
Aborted (core dumped)
```

I also make a summarize_graph for you, maybe it help

```
Found 1 possible inputs: (name=input_image, type=int32(3), shape=[1,?,?,3]) 
No variables spotted.
Found 2 possible outputs: (name=output_corner, op=Tile) (name=output_mask, op=Tile) 
Found 1471309 (1.47M) const parameters, 0 (0) variable parameters, and 51 control_edges
Op types used: 239 Const, 168 Switch, 36 FusedBatchNorm, 27 Add, 22 Identity, 21 Merge, 18 Conv2D, 15 Relu, 13 StridedSlice, 12 Shape, 7 Mul, 6 Cast, 5 MaxPool, 5 Pack, 5 RealDiv, 3 Slice, 3 Conv2DBackpropInput, 2 Tile, 2 Reshape, 2 RandomUniform, 2 PlaceholderWithDefault, 2 Greater, 2 Floor, 2 ConcatV2, 1 Placeholder, 1 Maximum, 1 LogicalOr, 1 ResizeBilinear, 1 Sigmoid, 1 Softmax, 1 Squeeze, 1 ExpandDims, 1 Sub, 1 ArgMax
```

**Comment by andrehentz** [comment]
TOCO is expecting a Switch node to be followed by a Merge node, but that doesn't seem to be the case in your model.

**Comment by Kayoku** [comment]
Hi, 
Thank you for your answer @andrehentz but I don't understand. You say that when I have a Switch node, I necessarily need to have a Merge node ?
So it's not good to have more Switch (168) than Merge (21) ?

For solve this, do I need to add manually a Merge node after every Switch node in my model ? I don't really know how to do that. Why TfLite don't do it by itself ?

Thx again for your time!

**Comment by adekunleba** [comment]
Hi Kayoku,

I encountered the same issue, would want to know how you later solved the issue.

Thanks

**Comment by Kayoku** [comment]
Hi @adekunleba ,
I didn't. I'm always stuck haha (I work on an other project waiting for a solution here). But if you find something, don't forget to tell me!
Sorry :(

**Comment by adekunleba** [comment]
Hi @Kayoku ,

Tried to check the source code and found out it's one of the assertions causing the error. 
```
// There remains to handle the edges that were pointing to the nonselected
  // output. We will just discard those edges. Concretely, at the moment,
  // our only examples of graphs with Switch nodes have them feeding into Merge
  // nodes, so what we're saying here is that we'll make the convention,
  // in our toco internal representation, that Merge nodes with only 1 input
  // are Merge nodes that have been resolved already and should be have as
  // Identity nodes, simply forwarding their input.
  //
  for (const auto& other_op : model->operators) {
    auto input_it = other_op->inputs.begin();
    while (input_it != other_op->inputs.end()) {
      if (*input_it == switch_op->outputs[nonselected_output_index]) {
        // Let us guard our assumption that only Merge nodes consume the outputs
        // of Switch nodes:
        CHECK(other_op->type == OperatorType::kTensorFlowMerge);  //ASSERTION CAUSING THE BOUNCE
        input_it = other_op->inputs.erase(input_it);
      } else {
        ++input_it;
      }
    }
  }

```
One approach I have tried is to comment that out the line, but seems it's an important check for another codeline, hence it's required to pass. For now, I worked using the tfmobile approach instead, as I look futher to how my architecture can be refactored to meet the merge requirements.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by farhan333** [comment]
Getting the same fail. Using LSTM.

2018-06-01 21:14:40.418639: F tensorflow/contrib/lite/toco/graph_transformations/resolve_tensorflow_switch.cc:95] Check failed: other_op->type == OperatorType::kTensorFlowMerge

**Comment by AManWithAPlan** [comment]
Hello, getting the same fail with next models: "ssd_inception_v2_coco_2017_11_17", "ssd_mobilenet_v1_coco_2017_11_17", "ssd_mobilenet_v2_coco_2018_03_29" .

2018-06-06 14:08:10.450529: I tensorflow/contrib/lite/toco/graph_transformations=/graph_transformations.cc:39] Before Removing unused ops: 4331 operators, 7153 arrays (0 quantized)
2018-06-06 14:08:10.776917: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 4276 operators, 7046 arrays (0 quantized)
2018-06-06 14:08:11.215557: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 4276 operators, 7046 arrays (0 quantized)
2018-06-06 14:08:11.639476: F tensorflow/contrib/lite/toco/graph_transformations/resolve_tensorflow_switch.cc:95] Check failed: other_op->type == OperatorType::kTensorFlowMerge

**Comment by xiadeye** [comment]
Any solutions?

**Comment by sanoguch** [comment]
I saw the same issue. Our models are created by keras.
It looks TOCO assumes that merge node is immediately after switch node, but in our models, there are several nodes between switch and merge.

I fixed as below and it works for our models. I am not sure it also works for others' models.

[resolve_tensorflow_switch.patch.zip](https://github.com/tensorflow/tensorflow/files/2097006/resolve_tensorflow_switch.patch.zip)

**Comment by ashermancinelli** [comment]
I am running into the same problem with a retrained Inception V3 model. Seems like Google's own model should be able to be converted into TFLite. Hoping a solution comes from this. [This](https://github.com/tf-coreml/tf-coreml/issues/200) also happens when trying to convert from frozen pb to `.mlmodel`.

**Comment by div1090** [comment]
I'm trying to convert a custom model on inception V3 + SSD and running into the same issue. Any workarounds to get the model to tflite format until this is officially fixed?

**Comment by ohadlights** [comment]
Same issue here... fail to convert a pretty simple mobile network.

**Comment by david-haber** [comment]
I have this issue when I convert networks with a placeholder as input:

`# Make placeholder to define phase
is_training = tf.placeholder_with_default(False, shape=(), name="is_training")`

Replacing this with

`is_training=False`

for inference lets me convert the model successfully. Making this manual change is not ideal though.

**Comment by ohadlights** [comment]
@david-haber your w/a worked for me

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by miaout17** [comment]
TensorFlow Lite doesn't support control flow ops yet (e.g. Switch, Merge, Cond...etc). The team is looking into broadening the support.

It's not possible to make Switch/Merge work properly with dynamic RNNs.

TFLite already implemented fused RNN / LSTM ops. However, there is no easy way to produce these TFLite ops through our converter. The team is currently working on a better way to convert these ops. Stay tuned.

====

> TOCO is expecting a Switch node to be followed by a Merge node, but that doesn't seem to be the case in your model.

In addition, TOCO can only resolve Switch/Merge in a very specific case (the predicate of Switch need to be constant). 
Therefore dynamic RNN won't work, because it requires real control flow when running the model. 

> I was looking to the file "tensorflow/contrib/lite/toco/model.h" for find "kTensorflowMerge", but it doesn't even exist... :D

It seems kTensorflowMerge was renamed to kMerge.

**Comment by arpitkh96** [comment]
Anyone got a workaround for this? I am having this problem with mobilenet v2 keras

**Comment by gargn** [comment]
@arpitkh96 The recommended option is to avoid using Merge in your graph. As far as I know, MobileNetV2 (classification) works with TFLite. Can you file a separate bug with the reproducible steps that are leading to this error. Include the command you are using for converting the model and attach the model.

I am closing the original bug due to lack of recent activity in relation to the original user's problem. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by arpitkh96** [comment]
@gargn This should suffice. https://github.com/arpitkh96/KerasMobilnet/blob/master/Untitled.ipynb
I'll try to open separate issue when I'll have time

**Comment by gargn** [comment]
@arpitkh96 One quick suggestion is to make sure that you are not passing in a training model by setting `keras.backend.set_learning_phase(False)`.

Additionally, `TocoConverter` has a `from_keras_model_file` function that makes it easier to do what you are doing, so I suggest using that. It is broken in 1.10, however, it has been fixed on the tf-nightly and in the 1.11 release.

**Comment by arpitkh96** [comment]
@gargn  learning phase and tf nightly finally fixed it. Thanks. BTW I was not loading from file because relu6 isn't recognized even after using CustomObjectScope

**Comment by david-haber** [comment]
I don't think this issue has been resolved yet. The current version on master does not allow me to convert models with placeholder_with_default ops in the graph.

> TOCO is expecting a Switch node to be followed by a Merge node, but that doesn't seem to be the case in your model.

@andrehentz Do we still want to impose this restriction? Otherwise I could implement a fix that allows us to convert placeholder_with_default ops.

**Comment by andrehentz** [comment]
I think there is partial support in TOCO for converting placeholder_with_default. The SwitchMerge restriction has to do with TF Lite not supporting control flow, right?

**Comment by JitenderMaurya** [comment]
initially i was getting this error " Check failed: IsConstantParameterArray(*model, bn_op->inputs[1]) && IsConstantParameterArray(*model, bn_op->inputs[2]) && IsConstantParameterArray(*model, bn_op->inputs[3]) Batch normalization resolution requires that mean, multiplier and offset arrays be constant." then i changed is_training = tf.placeholder_with_default(False, shape=(), name = "batchnormswitch") . since then i am getting " tensorflow/contrib/lite/toco/graph_transformations/resolve_tensorflow_switch.cc:95] Check failed: other_op->type == OperatorType::kTensorFlowMerge 
Aborted (core dumped)

None"

error Kindly help. Thanks in advance.

**Comment by farhan333** [comment]
Anybody have any luck with this?

**Comment by Samveed** [comment]
@sanoguch Where exactly should we add the patch you have given? Which files specifically?

**Comment by sanoguch** [comment]
I think it is tensorflow/lite/toco/graph_transformations/resolve_tensorflow_switch.cc could be replaced with this one.

[resolve_tensorflow_switch.zip](https://github.com/tensorflow/tensorflow/files/3210585/resolve_tensorflow_switch.zip)

Actually we gave up using TOCO or even TensorFlow Lite, so I am not actively following this issue. I am not sure it works or not.

**Comment by hamlatzis** [comment]
After applying the patch (https://github.com/tensorflow/tensorflow/issues/18731#issuecomment-396843328) I get the error below:

```
TOCO failed. See console for info.
2019-09-02 16:05:31.485705: F tensorflow/lite/toco/graph_transformations/propagate_array_data_types.cc:257] Check failed: op->inputs.size() > 0 (0 vs. 0)
Aborted (core dumped)
```

**Comment by IvanChavR** [comment]
Has someone have resolve this issue?, I want to convert the ssd_resnet50_v1 and ssd_inception_v2_coco and both of them throw **the OperatorType::kMerge Found StridedSlice as non-selected output from Switch, but only Merge supported.**
